ext/socket/lib/socket.rb

```diff
@@ -599,6 +599,47 @@ def accept_nonblock(exception: true)
     __accept_nonblock(exception)
   end
 
+  RESOLUTION_DELAY = 0.05
+  private_constant :RESOLUTION_DELAY
+
+  PATIENTLY_RESOLUTION_DELAY = 3
+  private_constant :PATIENTLY_RESOLUTION_DELAY
+
+  CONNECTION_ATTEMPT_DELAY = 0.25
+  private_constant :CONNECTION_ATTEMPT_DELAY
+
+  ADDRESS_FAMILIES = {
+    ipv6: Socket::AF_INET6,
+    ipv4: Socket::AF_INET
+  }.freeze
+  private_constant :ADDRESS_FAMILIES
+
+  HOSTNAME_RESOLUTION_QUEUE_UPDATED = 0
+  private_constant :HOSTNAME_RESOLUTION_QUEUE_UPDATED
+
+  @tcp_fast_fallback = true
+
+  class << self
+    attr_accessor :tcp_fast_fallback
+
+    def initialize_socket
+      @local_ip_address_list = Socket.ip_address_list
```

Comment:
I suppose caching `Socket.ip_address_list` permanently is not a good idea. The host may later gain or lose IP addresses on network configuration changes (e.g., Wi-Fi / cellular network) or when the OS is booting and the network interfaces are not yet ready.

Reply:
Thank you for your comment.
As you mentioned, persisting the cache could lead to unexpected issues, so we have decided to remove it. https://github.com/ruby/ruby/pull/9374/commits/a3d7056591c1f798b13893ec4ab603623980a7e2

Reply:
I'd also support removing the dual-/single-stack check because it could result in false negatives in certain corner cases, for example,
- When the host only has a link-local address, and /etc/hosts or mDNS allows resolving names of peer hosts to link-local addresses. In this case, the hosts can communicate using link-local addresses (APIPA network).
- When the host doesn't have an IPv6 GUA, and tries to connect to a hostname that resolves to ::1 (e.g., "localhost6" in /etc/hosts). In this case a PF_INET6 socket should be used.

I think it's difficult to enumerate this kind of unusual network configuration in the wild.